### PR TITLE
Fix `to_string_in_format_args` false positive

### LINKED ISSUE #9256: Incorrect suggested removal of to_string on slice
### Summary

`clippy::to_string_in_format_args` gives incorrect output when you call to_string on a str slice.
The code it suggests doesn't compile.

### Reproducer

I tried this code:

```rust
fn print_substring(original: &str) {
    assert!(original.len() > 10);
    println!("{}", original[..10].to_string());
}

fn main() {
    print_substring("Hello, world!");
}
```

I expected to see this happen:
Clippy would either produce no error, or offer a valid correction

Instead, this happened:
It offered this alternative, which doesn't compile.
original[..10] doesn't have a known size, so can't be passed to println!
It should be &original[..10]
```
warning: `to_string` applied to a type that implements `Display` in `println!` args
 --> src/main.rs:3:34
  |
3 |     println!("{}", original[..10].to_string());
  |                                  ^^^^^^^^^^^^ help: remove this
  ```

### Version

```text
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.62.0
LLVM version: 14.0.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

### PR DESCRIPTION
**Pull request by smoelius**
Fix #9256

changelog: none

### CODE REVIEW

File: clippy_lints/src/format_args.rs [hunk 1]
```diff
@@ -134,34 +134,49 @@ fn check_to_string_in_format_args(cx: &LateContext<'_>, name: Symbol, value: &Ex
...
-                        format!("{:*>width$}{}", "", receiver_snippet, width = n_needed_derefs),
-                        Applicability::MachineApplicable,
-                    );
-                }
+            let needed_ref = if implements_trait(cx, receiver_ty, sized_trait_id, &[]) {
+                ""
+            } else {
+                "&"
+            };
+            if n_needed_derefs == 0 && needed_ref.is_empty() {
```
Comment (llogiq) [inline clippy_lints/src/format_args.rs:148]: How about a `needs_ref` instead of a `needed_ref` so we don't need an `is_empty()` here?

**Review by llogiq — APPROVED** [review]
Just one small nit pick, otherwise this looks good.

**Comment by smoelius** [comment]
Hopefully, I understood correctly. Thanks for reviewing, @llogiq.

**Comment by llogiq** [comment]
Yes, that's what I meant. Sorry for being absent for a while. r=me after rebase.

**Comment by smoelius** [comment]
Rebase done. Thanks again, @llogiq.

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Pull request merged**